### ISSUE TITLE
FFTW3 linker error

### ISSUE THREAD

**Issue by danieltuzes** [microsoft/vcpkg#6834]
I run `.\vcpkg.exe install fftw3:x64-windows` successfully. I added `#include <fftw3.h>` to my code and tried to compile. I got the errors
```
error LNK2001: unresolved external symbol __imp_fftw_plan_dft_r2c_1d
error LNK2001: unresolved external symbol __imp_fftw_execute
```
in release mode, and in debug mode i got:

```
error LNK2019: unresolved external symbol __imp_fftw_execute referenced in function main
error LNK2019: unresolved external symbol __imp_fftw_plan_dft_r2c_1d referenced in function main
```
I already run `.\vcpkg.exe integrate install` successfully too, before this install. Did I do something wrong? In folder `vcpkg\packages\fftw3_x64-windows` I didn't find any .dll file.

I am using MSVS Community 2019 version 16.1.2 and vcpkg version 2018.11.23-nohash.

**Comment by danieltuzes** [comment]
I managed to use it by installing it manually.
* download the windows dlls from (FFTW)[http://www.fftw.org/install/windows.html]
* run the lib.exe on the files as described on the page
* copied the files near my project and added that folder as an Additional Library Directory for the Linker
* added `libfftw3-3.lib;libfftw3f-3.lib;libfftw3l-3.lib` as Additional Dependencies

**Comment by Neumann-A** [comment]
> Important: If you want to compile FFTW as a DLL, you should add a line #define FFTW_DLL to fftw3.h and ifftw.h before compiling. This will add the requisite __declspec decorations to function declarations etc. 

https://github.com/microsoft/vcpkg/blob/a6960ed0b168a06414fce1e4063772134f319f3d/ports/fftw3/portfile.cmake#L102-L108

SOURCE_PATH must be replaced by CURRENT_PACKAGES_DIR

as a workaround add FFTW_DLL to your preprocessor definitions. For CMake users we could patch fftw3 CMakeLists where `add_definition(-DFFTW_DLL) ` could be replaced with a `target_compile_defintions(... PUBLIC FFTW_DLL) `call which than adds the defintion to the generated targets. Too late in my timezone for doing that now ;)

**Comment by danieltuzes** [comment]
Dear @Neumann-A 
Thank you for your fast reply.
1. I am sorry but I am not sure what your answer is referring to exactly. After the above mentioned handmade installation I still needed to provide the dll-s, but this wasn't my problem. My problem was that although I think I followed the procedure of the fftw installation via vcpkg, I still couldn't compile my code due to a linker error (and not runtime error). Was your answer referring to this issue?
2. My fftw.h contains `#define FFTW_DLL` at line 50 (the 3rd non commented line). However, ifftw.h cannot be found.
3. A took a look on `vcpkg/ports/fftw3/portfile.cmake` but I couldn't find the lines mentioned. Was  "SOURCE_PATH must be replaced by CURRENT_PACKAGES_DIR" an instruction for me?
4. I added `FFTW_DLL` to the preprocessor definitions. After that I got the same errors and a new warning: `C:\local\vcpkg\installed\x64-windows\include\fftw3.h(50,1): warning C4005:  'FFTW_DLL': macro redefinition`.

**Comment by Neumann-A** [comment]
> I am sorry but I am not sure what your answer is referring to exactly. 

Interesting what github shows and what can be checked out with git are too different things... 

the DLLs should be located at `vcpkg\packages\fftw3_x64-windows\bin` or `vcpkg\installed\x64-windows\bin`

```
error LNK2001: unresolved external symbol __imp_fftw_plan_dft_r2c_1d
error LNK2001: unresolved external symbol __imp_fftw_execute
```

I cannot find those symbols in fftw3. Do you have other macro redefinition warnings in your build? I also checked how fftw3 generates its name and this does not look like one of those generated names. They should all start with `fftw<prec>_` .
as seen here:

```
#define FFTW_CONCAT(prefix, name) prefix ## name
#define FFTW_MANGLE_DOUBLE(name) FFTW_CONCAT(fftw_, name)

#define FFTW_DEFINE_API(X, R, C)                                        \
FFTW_EXTERN void                                                        \
FFTW_CDECL X(execute)(const X(plan) p);                                 

#if defined(FFTW_DLL) && (defined(_WIN32) || defined(__WIN32__))
   /* annoying Windows syntax for shared-library declarations */
#  if defined(COMPILING_FFTW) /* defined in api.h when compiling FFTW */
#    define FFTW_EXTERN extern __declspec(dllexport)
#  else /* user is calling FFTW; import symbol */
#    define FFTW_EXTERN extern __declspec(dllimport)
#  endif
#else
#  define FFTW_EXTERN extern
#endif

#if defined(_WIN32) || defined(__WIN32__)
#  define FFTW_CDECL __cdecl
#else
#  define FFTW_CDECL
#endif

FFTW_DEFINE_API(FFTW_MANGLE_DOUBLE, double, fftw_complex)
```

Maybe try again with a clean VCPKG folder?

**Comment by danieltuzes** [comment]
Thank you @Neumann-A again for the fast reply.
1. No fftw dlls were found `vcpkg\packages\fftw3_x64-windows\bin` or `vcpkg\installed\x64-windows\bin`. Especially, the first folder doesn't even exist. 
2. Only those 2 errors were present, no other warnings were shown. 
3. What do you mean by "a clean vcpkg folder"? I created a new folder, pulled vcpkg there, installed with `.\bootstrap-vcpkg.bat` and asked `.\vcpkg install fftw3:x64-windows` but I was told that
```
The following packages are already installed:
    fftw3[core]:x64-windows
Starting package 1/1: fftw3:x64-windows
Package fftw3:x64-windows is already installed
```
And of course, point 3. didn't solve my problem.

After these I tried the following as well without any success. I restarted VS and created a new solution and project that requires boost and fftw (both installed with vcpkg). But I have got the same linker errors again, only for fftw.

One more thing to mention: if I add `/VERBOSE` to the linker, it tells that
```
1>    Searching C:\local\vcpkg\installed\x64-windows\lib\fftw3.lib:
1>    Searching C:\local\vcpkg\installed\x64-windows\lib\fftw3f.lib:
1>    Searching C:\local\vcpkg\installed\x64-windows\lib\fftw3l.lib:
```
and they are indeed there, but nothing else in regards to fftw (beside the errors). Shouldn't it tell that it found something in those libraries? Those libs in vcpkg's folder are considerably larger (9MB) than those I produced from the downloaded file from FFTW's webpage (200KB, but with those come the dlls at around 2MB). NB: the self-produced libs are named like `libfftw3-3.lib` are added to as an additional library to the linker in the project properties.

**Comment by Neumann-A** [comment]
> No fftw dlls were found vcpkg\packages\fftw3_x64-windows\bin or vcpkg\installed\x64-windows\bin. Especially, the first folder doesn't even exist.

that is strange

> What do you mean by "a clean vcpkg folder"? I created a new folder, pulled vcpkg there, installed with .\bootstrap-vcpkg.bat and asked .\vcpkg install fftw3:x64-windows but I was told that

that is even stranger. Are you sure the folder was empty?

>After these I tried the following as well without any success. I restarted VS and created a new solution and project that requires boost and fftw (both installed with vcpkg). But I have got the same linker errors again, only for fftw.

has nothing to do with VS so anything you do there will do nothing to solve the problem.
You VCPKG folder/installation is somehow corrupted.

> Shouldn't it tell that it found something in those libraries? Those libs in vcpkg's folder are considerably larger (9MB) than those I produced from the downloaded file from FFTW's webpage 

That tells me you are using an old VCPKG version where FFTW was building a static library using x64-windows. (which was changed 3 weeks ago or so)

My best guess here is delete anything vcpkg related and remove the vs integration first. Then restart from scratch with `git clone https://github.com/microsoft/vcpkg.git`

**Comment by danieltuzes** [comment]
Dear @Neumann-A 
Thank you for your efforts, we succeed. I removed the vcpkg folder and recreated, then pulled vcpkg and installed again all the packages. It works now.

Was the vcpkg faulty, or was the package information too old? How could I refresh the required units next time without removing the whole folder and reinstalling all the packages?

And a wish: is there a way to ask the compiler to merge the `fftw3.dll` file with my executable in a single file so I don't need to copy both of them to the place where I'd like to run them?
